src/bosh_azure_cpi/lib/cloud/azure/stemcell/stemcell_manager2.rb

```diff
@@ -50,12 +96,142 @@ def has_stemcell?(storage_account_name, name)
 
     def get_user_image_info(stemcell_name, storage_account_type, location)
       @logger.info("get_user_image_info(#{stemcell_name}, #{storage_account_type}, #{location})")
+
+      if _compute_gallery_enabled?
+        @logger.debug("Compute gallery is enabled. Try to get image in '#{location}' for stemcell '#{stemcell_name}'")
+        gallery_image = _ensure_gallery_image_in_target_location(stemcell_name, location)
+
+        unless gallery_image.nil?
+          @logger.debug("Using gallery image: #{gallery_image[:id]}")
+          return StemcellInfo.new(gallery_image[:id], gallery_image[:tags])
+        end
+      end
+
+      @logger.debug("Try to get user image in '#{location}' for stemcell '#{stemcell_name}'")
       user_image = _get_user_image(stemcell_name, storage_account_type, location)
-      StemcellInfo.new(user_image[:id], user_image[:tags])
+      return StemcellInfo.new(user_image[:id], user_image[:tags])
     end
 
     private
 
+    def _compute_gallery_enabled?
+      !@azure_config.compute_gallery_name.nil? && !@azure_config.compute_gallery_name.empty?
+    end
+
+    def _make_semver_compliant(version)
+      v = version.split('.')
+      major = v[0]
+      minor = v.length > 1 ? v[1] : 0
+      patch = v.length > 2 ? v[2] : 0
+      "#{major}.#{minor}.#{patch}"
+    end
+
+    def _create_gallery_image(stemcell_name, image_definition, version, location, metadata)
+      gallery_name = @azure_config.compute_gallery_name
+      os_type = metadata['os_type']&.downcase&.capitalize
+      cloud_error("Invalid os_type '#{os_type}'") unless ['Linux', 'Windows'].include?(os_type)
+
+      params = { 'location' => location, 'osType' => os_type }.merge(JSON.parse(metadata['image']))
+      flock("#{CPI_LOCK_CREATE_GALLERY_IMAGE}-#{gallery_name}-#{image_definition}", File::LOCK_EX) do
+        @logger.debug("Ensuring compute gallery image definition '#{gallery_name}/#{image_definition}'")
+        @azure_client.create_gallery_image_definition(gallery_name, image_definition, params)
+      end
+
+      gallery_image = nil
+      flock("#{CPI_LOCK_CREATE_GALLERY_IMAGE}-#{gallery_name}-#{image_definition}-#{version}", File::LOCK_EX) do
+        @logger.debug("Creating compute gallery image '#{gallery_name}/#{image_definition}:#{version}' in target location '#{location}'")
+        metadata['stemcell_name'] = stemcell_name
+        gallery_image = @azure_client.create_gallery_image_version(
+          gallery_name,
+          image_definition,
+          version,
+          {
+            'location'             => location,
+            'tags'                 => metadata,
+            'storage_account_name' => @default_storage_account_name,
```

Comment:
ℹ️ The usage of a storage account here confused me initially, but it appears even when using managed disks the CPI will either search for a usable storage account, or create one if it can't find one.

Reply:
Yeah, I also needed a moment to understand this. CPI will then attempt to copy the stemell blob into that storage account.